qupulse/program/linspace.py

```diff
@@ -24,45 +146,108 @@ class DepKey:
-        return cls(tuple(int(round(voltage / resolution)) for voltage in voltages))
+        while factors and factors[-1] == 0:
+            factors = factors[:-1]
+        return cls(tuple(int(round(factor / resolution)) for factor in factors),
+                   domain)
+    
+    @classmethod
+    def from_voltages(cls, voltages: Sequence[float], resolution: float):
+        return cls.from_domain(voltages, resolution, DepDomain.VOLTAGE)
+    
+    @classmethod
+    def from_lin_times(cls, times: Sequence[float], resolution: float):
+        return cls.from_domain(times, resolution, DepDomain.TIME_LIN)
 
 
 @dataclass
 class LinSpaceNode:
     """AST node for a program that supports linear spacing of set points as well as nested sequencing and repetitions"""
-
-    def dependencies(self) -> Mapping[int, set]:
+        
+    def dependencies(self) -> Mapping[GeneralizedChannel, set]:
+        # doing this as a set _should_ get rid of non-active deps that are one level above?
+        #!!! 
         raise NotImplementedError
 
+@dataclass
+class LinSpaceNodeChannelSpecific(LinSpaceNode):
+    
+    channels: Tuple[GeneralizedChannel, ...]
+    
+    @property
+    def play_channels(self) -> Tuple[ChannelID, ...]:
+        return tuple(ch for ch in self.channels if isinstance(ch,ChannelID))
+    
 
 @dataclass
-class LinSpaceHold(LinSpaceNode):
+class LinSpaceHold(LinSpaceNodeChannelSpecific):
     """Hold voltages for a given time. The voltages and the time may depend on the iteration index."""
 
-    bases: Tuple[float, ...]
-    factors: Tuple[Optional[Tuple[float, ...]], ...]
+    bases: Dict[GeneralizedChannel, float]
+    factors: Dict[GeneralizedChannel, Optional[Tuple[float, ...]]]
 
     duration_base: TimeType
     duration_factors: Optional[Tuple[TimeType, ...]]
-
-    def dependencies(self) -> Mapping[int, set]:
-        return {idx: {factors}
-                for idx, factors in enumerate(self.factors)
+    
+    def dependencies(self) -> Mapping[DepDomain, Mapping[ChannelID, set]]:
+        return {dom: {ch: {factors}}
+                for dom, ch_to_factors in self._dep_by_domain().items()
+                for ch, factors in ch_to_factors.items()
                 if factors}
-
+    
+    def _dep_by_domain(self) -> Mapping[DepDomain, Mapping[GeneralizedChannel, set]]:
+        return {DepDomain.VOLTAGE: self.factors,
+                DepDomain.TIME_LIN: {DepDomain.TIME_LIN:self.duration_factors},
+                }
+    
 
 @dataclass
-class LinSpaceArbitraryWaveform(LinSpaceNode):
+class LinSpaceArbitraryWaveform(LinSpaceNodeChannelSpecific):
     """This is just a wrapper to pipe arbitrary waveforms through the system."""
     waveform: Waveform
-    channels: Tuple[ChannelID, ...]
+    
+    def dependencies(self):
+        return {}
+    
 
+@dataclass
+class LinSpaceArbitraryWaveformIndexed(LinSpaceNodeChannelSpecific):
+    """This is just a wrapper to pipe arbitrary waveforms through the system."""
```

Comment:
I guess this is no longer true

Reply:
yes, docstrings are of course not adjusted atm